### ISSUE TITLE
Use sucker_punch instead of Thread.new for asynchronous processes

### ISSUE THREAD

**Issue by ryanfb** [sosol/sosol#71]
Asynchronous processes (e.g. `Vote#tally_votes` and `PublicationsController#become_finalizer`) currently use `Thread.new` to get asynchronicity. This requires a lot of work in the tests to make sure we wait on Threads, as well as a lot of custom work around the Thread to make sure we set up and terminate e.g. the ActiveRecord + Rails.logger instances it inherits.

I've looked into [other gems for asynchronous processing](https://www.ruby-toolbox.com/categories/Background_Jobs) in the past, but most of these require a separate process to run the job queue, and many require e.g. a Redis store for the job queue itself.

In contrast, [sucker_punch](https://github.com/brandonhilkert/sucker_punch) runs inside the web process, and [the queue pooling mechanism it's built off of is designed for JRuby](https://github.com/celluloid/celluloid#supported-platforms). See also: http://brandonhilkert.com/blog/why-i-wrote-the-sucker-punch-gem/

Converting the existing `Thread.new` code to Sucker Punch jobs wasn't too difficult, and made testing much cleaner and easier (though it currently avoids async testing): https://github.com/sosol/sosol/tree/sucker_punch https://travis-ci.org/sosol/sosol/builds/70968133

The only downside to this compared to some other delayed job gems is that if the server process is terminated while a job is running/queued, it won't be resumed when the server is restarted. But we would already have that problem with `Thread.new`, and it's probably easier to work around with `sucker_punch` if we eventually need to.

Thoughts on integrating this and making sure it works as expected inside Tomcat on the next dev testing round, @hcayless?

**Comment by hcayless** [comment]
Yeah, I've no problem at all with this. Using Thread,new was just the most
obvious and simple way to see if things would work. Having a job
queue/transaction log might be something we do in the future, but I think
that would mean a major overhaul and therefore is out of scope right now.
Let's do it!
